src/Security/Authentication/Certificate/src/CertificateAuthenticationHandler.cs

```diff
@@ -0,0 +1,269 @@
+        {
+            // You only get client certificates over HTTPS
+            if (!Context.Request.IsHttps)
+            {
+                return AuthenticateResult.NoResult();
+            }
+
+            var clientCertificate = await Context.Connection.GetClientCertificateAsync();
+
+            // This should never be the case, as cert authentication happens long before ASP.NET kicks in.
+            if (clientCertificate == null)
+            {
+                Logger.LogDebug("No client certificate found.");
+                return AuthenticateResult.NoResult();
+            }
+
+            // If we have a self signed cert, and they're not allowed, exit early and not bother with
+            // any other validations.
+            if (clientCertificate.IsSelfSigned() &&
+                !Options.AllowedCertificateTypes.HasFlag(CertificateTypes.SelfSigned))
+            {
+                Logger.LogWarning("Self signed certificate rejected, subject was {0}", clientCertificate.Subject);
+
+                return AuthenticateResult.Fail("Options do not allow self signed certificates.");
+            }
+
+            // If we have a chained cert, and they're not allowed, exit early and not bother with
+            // any other validations.
+            if (!clientCertificate.IsSelfSigned() &&
+                !Options.AllowedCertificateTypes.HasFlag(CertificateTypes.Chained))
+            {
+                Logger.LogWarning("Chained certificate rejected, subject was {0}", clientCertificate.Subject);
+
+                return AuthenticateResult.Fail("Options do not allow chained certificates.");
+            }
+
+            var chainPolicy = BuildChainPolicy(clientCertificate);
+
+            try
+            {
+                var chain = new X509Chain
+                {
+                    ChainPolicy = chainPolicy
+                };
+
+                var certificateIsValid = chain.Build(clientCertificate);
+
+                if (!certificateIsValid)
+                {
+                    using (Logger.BeginScope(clientCertificate.GetCertHashString(HashAlgorithmName.SHA256)))
+                    {
+                        Logger.LogWarning("Client certificate failed validation, subject was {0}", clientCertificate.Subject);
+                        foreach (var validationFailure in chain.ChainStatus)
+                        {
+                            Logger.LogWarning("{0} {1}", validationFailure.Status, validationFailure.StatusInformation);
+                        }
+                    }
+                    return AuthenticateResult.Fail("Client certificate failed validation.");
+                }
+
+                var validateCertificateContext = new ValidateCertificateContext(Context, Scheme, Options)
+                {
+                    ClientCertificate = clientCertificate
+                };
+
+                await Events.ValidateCertificate(validateCertificateContext);
```

Comment:
CertificateValidated

In the other providers we build the Principal before calling this success event and pass it in. Then the dev can customize the identity without creating it from scratch.
https://github.com/aspnet/AspNetCore/blob/4e44025a52e4b73aa17e09a8041b0e166e0c5ce0/src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs#L598-L604

Reply:
I don't think that's the intent here, @blowdart can confirm, but it looks like this event is meant to be used to do more complicated custom validation, as opposed to a 'Success' event

Reply:
It's a bit of both, but yea, mostly custom validation. Weird we do both. Could move the call to CreatePrincipal from line 124 up, then you get the default principal that you can fuck around with in the event

Reply:
It's run after the built in validation has already passed, we think it's valid. If you wanted to replace the validation you'd need an event at line 62.

Reply:
Ok renamed